presto-main/src/main/java/com/facebook/presto/catalogserver/RemoteMetadataManager.java

```diff
@@ -0,0 +1,780 @@
+import io.airlift.slice.Slice;
+
+import javax.inject.Inject;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+// TODO : Use thrift to serialize metadata objects instead of json serde on catalog server in the future
+public class RemoteMetadataManager
+        implements Metadata
+{
+    // This value represents an empty result from the catalog server
+    private static final String EMPTY_STRING = "";
+
+    private final Metadata delegate;
+    private final TransactionManager transactionManager;
+    private final ObjectMapper objectMapper;
+    private final DriftClient<CatalogServerClient> catalogServerClient;
+
+    @Inject
+    public RemoteMetadataManager(
+            MetadataManager metadataManager,
+            TransactionManager transactionManager,
+            ObjectMapper objectMapper,
+            DriftClient<CatalogServerClient> catalogServerClient)
+    {
+        this.delegate = requireNonNull(metadataManager, "metadata is null");
+        this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+        this.objectMapper = requireNonNull(objectMapper, "objectMapper is null");
+        this.catalogServerClient = requireNonNull(catalogServerClient, "catalogServerClient is null");
+    }
+
+    @Override
+    public final void verifyComparableOrderableContract()
+    {
+        delegate.verifyComparableOrderableContract();
+    }
+
+    @Override
+    public Type getType(TypeSignature signature)
+    {
+        return delegate.getType(signature);
+    }
+
+    @Override
+    public void registerBuiltInFunctions(List<? extends SqlFunction> functionInfos)
+    {
+        delegate.registerBuiltInFunctions(functionInfos);
+    }
+
+    @Override
+    public boolean schemaExists(Session session, CatalogSchemaName schema)
+    {
+        return catalogServerClient.get().schemaExists(
+                transactionManager.getTransactionInfo(session.getRequiredTransactionId()),
+                session.toSessionRepresentation(),
+                schema);
+    }
+
+    @Override
+    public boolean catalogExists(Session session, String catalogName)
+    {
+        return catalogServerClient.get().catalogExists(
+                transactionManager.getTransactionInfo(session.getRequiredTransactionId()),
+                session.toSessionRepresentation(),
+                catalogName);
+    }
+
+    @Override
+    public List<String> listSchemaNames(Session session, String catalogName)
+    {
+        String schemaNamesJson = catalogServerClient.get().listSchemaNames(
+                transactionManager.getTransactionInfo(session.getRequiredTransactionId()),
+                session.toSessionRepresentation(),
+                catalogName);
+        if (!schemaNamesJson.equals(EMPTY_STRING)) {
```

Comment:
schemaNamesJson.isEmpty() ?

Reply:
As a string value is currently returned by the catalog server isEmpty isnt supported. Though I will say that once we move to using Thrift transport instead of JSON serde in the future using string checks wont be used any more and we will have a thrift struct returned instead

Reply:
Not sure if I follow, I was suggesting we can use isEmpty method instead of current mechanism.

Reply:
Oh my bad, I think I misinterpreted it as in you wanted me to use a list instead. Thanks for the catch, applied